src/NuGet.Core/NuGet.Commands/TrustedSignersCommand/TrustedSignerActionsProvider.cs

```diff
@@ -0,0 +1,310 @@
+                {
+                    reposig = primarySignature as RepositoryPrimarySignature;
+                }
+                else
+                {
+                    var countersignature = RepositoryCountersignature.GetRepositoryCountersignature(primarySignature);
+                    reposig = countersignature ?? throw new InvalidOperationException(Strings.Error_RepoTrustExpectedRepoSignature);
+                }
+
+                v3ServiceIndex = reposig.V3ServiceIndexUrl.AbsoluteUri;
+            }
+
+            var signers = _trustedSignersProvider.GetTrustedSigners();
+            foreach (var existingSigner in signers)
+            {
+                if (string.Equals(existingSigner.Name, name, StringComparison.Ordinal))
+                {
+                    throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, Strings.Error_TrustedSignerAlreadyExists, name));
+                }
+
+                if (trustingRepository && existingSigner is RepositoryItem repoItem && string.Equals(repoItem.ServiceIndex, v3ServiceIndex, StringComparison.OrdinalIgnoreCase))
+                {
+                    throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, Strings.Error_TrustedRepoAlreadyExists, v3ServiceIndex));
+                }
+            }
+
+            if (trustingRepository)
+            {
+                var certificateItem = GetCertificateItemForSignature(reposig, allowUntrustedRoot);
+
+                _trustedSignersProvider.AddOrUpdateTrustedSigner(new RepositoryItem(name, v3ServiceIndex, CreateOwnersList(owners), certificateItem));
+            }
+
+            if (trustTarget.HasFlag(VerificationTarget.Author))
+            {
+                if (primarySignature.Type != SignatureType.Author)
+                {
+                    throw new InvalidOperationException(Strings.Error_AuthorTrustExpectedAuthorSignature);
+                }
+
+                var certificateItem = GetCertificateItemForSignature(primarySignature, allowUntrustedRoot);
+
+                _trustedSignersProvider.AddOrUpdateTrustedSigner(new AuthorItem(name, certificateItem));
+            }
+        }
+
+#endif
+
+        /// <summary>
+        /// Adds a new trusted author to the settings.
+        /// If a trusted signer already exists with this name, adds a certificate item to it.
+        /// </summary>
+        /// <param name="name">Name of the trusted author.</param>
+        /// <param name="fingerprint">Fingerprint to be added to the certificate entry.</param>
+        /// <param name="hashAlgorithm">Hash algorithm used to calculate <paramref name="fingerprint"/>.</param>
+        /// <param name="allowUntrustedRoot">Specifies if allowUntrustedRoot should be set to true in the certificate entry.</param>
+        public void AddOrUpdateTrustedSigner(string name, string fingerprint, HashAlgorithmName hashAlgorithm, bool allowUntrustedRoot)
```

Comment:
Should the name of this api be `AddOrUpdateTrustedAuthorSigner` or `AddOrUpdateTrustedAuthor`?

Reply:
No, if the given name matches one of an existing signer (no matter if it's author or repository) this method will add the certificate to its list of certificates. Only if there exists no signer it will create an author item.

Reply:
then it makes it lil more confusing saying it updates both but only add for Author signer...

Reply:
The idea is that it will add a certificate to the list of a signer... If such signer doesn't exist it will default to create an author signer.... What do you suggest for naming?